File: todo_app/backend/src/middleware/auth.middleware.ts
import type { Request, Response, NextFunction } from "express";
import { verifyToken } from "../utils/token";

export const authMiddleware = (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  const header = req.headers.authorization;

  if (!header || !header.startsWith("Bearer ")) {
    return res.status(401).json({ error: "No token provided" });
  }

  const token = header.split(" ")[1];
  const payload = verifyToken(token as string);

  if (!payload) {
    return res.status(401).json({ error: "Invalid token" });
  }

  (req as any).userId = payload.userId;
  next();
};

File: freelance_market_place/backend/src/miiddleware/auth_middleware.py
from fastapi import Request
from fastapi.responses import JSONResponse
from src.utils.token import verify_token

async def auth_middleware(
    request: Request, 
    call_next
):
    allowed_paths = {"/auth/auth/login", "/auth/auth/signup"}
    if request.url.path in allowed_paths:
        return await call_next(request)
    
    token = request.headers.get("Authorization")

    if (not token):
        return JSONResponse({
            "message" : "No token found"
        }, status_code=401)

    if not token.startswith("Bearer "):
        return JSONResponse({
            "message" : "Invalid authorization header"
        }, 401)
     
    raw_token = token.split(" ")[1]

    data = verify_token(raw_token)
    if not data:
        return JSONResponse(
            {
                "details" : "Token not valid"
            }, 401
        )

    request.state.user = data
    return await call_next(request)
